### PR TITLE
Feature/chapter 1

### LINKED ISSUE #2: Chapter 1: Why version control
## Background

Write chapter that covers the following:

1. What benefits does version control have
2. Different ways of versioning your code manual -> git
3. Introduction to collaboration software inc git and vsts

## To do 

1. Create 01-whyversioncontrol.Rmd to make this chapter 2 in the bookdown.
2. Iterate guidance

### PR DESCRIPTION
**Pull request by adamrobinson361**
**Background**

To resolve #2 

**QA Required**

@calstaff91 , @ZacharyWaller  - can you both take a read through the preface and chapter 1 and provide feedback via this pull request on how you think it should be changed.

Note: @drmarkc13 added for proof reading too

### CODE REVIEW

**Comment by adamrobinson361** [comment]
Can you both indivudally comment before either approves as i'm not clear how to implement 2 reviews yet.

File: 01-intro.Rmd [hunk 1]
```diff
@@ -1,2 +1,79 @@
 # Introduction {#intro}
 
+## Background
+
+Version control is an important part of any project. It allows you to reactively respond to request for changes, ensure that you can easily reproduce an earlier iteration and provides an audit trail of what work has been done and the decisions that have been made.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: to respond reactively

Comment (drmarkc13) [reply 01-intro.Rmd]: to a request

or

to requests

File: 01-intro.Rmd [hunk 2]
```diff
@@ -1,2 +1,79 @@
 # Introduction {#intro}
 
+## Background
+
+Version control is an important part of any project. It allows you to reactively respond to request for changes, ensure that you can easily reproduce an earlier iteration and provides an audit trail of what work has been done and the decisions that have been made. 
+
+Most people would recognise this and probably are doing some form of their own version control to ensure they are able to, albeit not immediately, respond to these situations.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: able to respond, albeit not immediately, to...

File: 01-intro.Rmd [hunk 3]
```diff
@@ -1,2 +1,79 @@
 # Introduction {#intro}
 
+## Background
+
+Version control is an important part of any project. It allows you to reactively respond to request for changes, ensure that you can easily reproduce an earlier iteration and provides an audit trail of what work has been done and the decisions that have been made. 
+
+Most people would recognise this and probably are doing some form of their own version control to ensure they are able to, albeit not immediately, respond to these situations. 
+
+As analytical functions move away from software such as Excel, SPSS etc. towards fully code based solutions using languages such as R, Python, SQL etc. there is a wealth of pre-established expertise that can be lifted from traditional software development practices.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: SPSS, etc.

code-based

SQL, etc., there

File: 01-intro.Rmd [hunk 4]
```diff
@@ -1,2 +1,79 @@
...
+
+## Manual Version Control
+
+Any book illustrating the usefulness of git will start with an example similar to as follows.
+
+![](www/manual_version_control.png)
+
+This approach creates *versioned* files using suffixes with some sort of convention. This can work if you have a small number of iterations but the following questions need to be answered:
+
+1. **Which is the final version!** - Although conventions attempt to remain standard you will get to a stage when you think you are final, another change will come and your previous convention will be broken.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: Which is the final version?!

File: 01-intro.Rmd [hunk 5]
```diff
@@ -1,2 +1,79 @@
...
+
+![](www/manual_version_control.png)
+
+This approach creates *versioned* files using suffixes with some sort of convention. This can work if you have a small number of iterations but the following questions need to be answered:
+
+1. **Which is the final version!** - Although conventions attempt to remain standard you will get to a stage when you think you are final, another change will come and your previous convention will be broken. 
+
+2. **What files actually constitute this analysis?** - Folder structure changes over time. Scripts merge, are broken down into subscripts and files that may no longer be needed could be sitting in your folder. For a colleague quality assuring your work or anyone (including yourself) picking this up in the future, this structure is not going to help you as it will be unclear what is really needed. 
+
+3. **What are the differences between these versions?** - Although these versions have different numbers we cant easily tell what the difference between them are and would have to manually compare differences to see what's going on. There is no indication that these are linked to specific items of work.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: can't

File: 01-intro.Rmd [hunk 6]
```diff
@@ -1,2 +1,79 @@
...
+
+4. **How do you collaborate using this structure?** - Collaborating using this sort of versioning means you are going to have to introduce others to this structure and/ or manage the process of adding pieces of code, likely passed across via email. Keeping track of this adds to the challenge.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: and/or (no space?)

File: 01-intro.Rmd [hunk 7]
```diff
@@ -1,2 +1,79 @@
...
+
+The main take home points from this is that there are lots of challenges that come with this and these challenges can take away from the task at hand. The bigger your project grows, the more challenging this will be.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: 'are' not 'is'

File: 01-intro.Rmd [hunk 8]
```diff
@@ -1,2 +1,79 @@
...
+
+3. **What are the differences between these versions?** - Although these versions have different numbers we cant easily tell what the difference between them are and would have to manually compare differences to see what's going on. There is no indication that these are linked to specific items of work.
+
+4. **How do you collaborate using this structure?** - Collaborating using this sort of versioning means you are going to have to introduce others to this structure and/ or manage the process of adding pieces of code, likely passed across via email. Keeping track of this adds to the challenge.
+
+The main take home points from this is that there are lots of challenges that come with this and these challenges can take away from the task at hand. The bigger your project grows, the more challenging this will be. 
+
+## Git
+
+Git is the [most commonly used version control system](https://insights.stackoverflow.com/survey/2015#tech-sourcecontrol) for code development in the world today.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: Get hyperlink to open in new window?

File: 01-intro.Rmd [hunk 9]
```diff
@@ -1,2 +1,79 @@
...
+
+4. **How do you collaborate using this structure?** - Collaborating using this sort of versioning means you are going to have to introduce others to this structure and/ or manage the process of adding pieces of code, likely passed across via email. Keeping track of this adds to the challenge.
+
+The main take home points from this is that there are lots of challenges that come with this and these challenges can take away from the task at hand. The bigger your project grows, the more challenging this will be. 
+
+## Git
+
+Git is the [most commonly used version control system](https://insights.stackoverflow.com/survey/2015#tech-sourcecontrol) for code development in the world today. 
+
+It is command line software, language agnostic and at its route, just looks at differences between plain text files.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: 'root' not 'route'

File: 01-intro.Rmd [hunk 10]
```diff
@@ -1,2 +1,79 @@
...
+
+At its most basic level git allows the following:
+
+1. **Commits** - Create checkpoints in your code development process that you can go back to if required. 
+
+2. **Branch** - Create exact replicas of your project folder in which you can try out particular ideas without editing the main version. 
+
+3. **Merge** - Systematically merge these changes back into the main version of the code.
+
+Git works in a way such that each time you make a change it is not a new version of the code is stored but the differences between itself and the previous version.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: new version of the code that is stored...

File: 01-intro.Rmd [hunk 11]
```diff
@@ -1,2 +1,79 @@
...
+
+3. **Merge** - Systematically merge these changes back into the main version of the code.
+
+Git works in a way such that each time you make a change it is not a new version of the code is stored but the differences between itself and the previous version. 
+
+This is invaluable information and means that rather than checking the whole project code at an undetermined time in a project the task becomes:
+
+*What changed since the last working version, was that done right and does that change impact anywhere else?*.
+
+**Note:** Git is command line software but many development environment (including R Studio) provide the functionality via in the graphical user interface.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: development environments

via a graphical....

File: 01-intro.Rmd [hunk 12]
```diff
@@ -1,2 +1,79 @@
...
+
+What is needed is a visual platform where central versions of the code can be stored with the following functionality:
+
+1. Users can take copies of the code on their local machine 
+
+2. Users can make changes to the code locally and submit those changes for review
+
+3. The review process can be managed and if any changes are required they can be made in a systemised manner.
+
+You can think of these sort of systems as wrappers git where merges into the main version of the code are safeguarded.
```
Comment (drmarkc13) [inline 01-intro.Rmd]: wrappers for git

File: index.Rmd [hunk 13]
```diff
@@ -9,4 +9,24 @@ link-citations: yes
...
 
-# Preface {-}
\ No newline at end of file
+# Preface {-}
+
+## Overview
+
+This book aims act as a resource for analysts on how and why they should use version control.
+
+It is in the context of the Department for Education Corporate environment and focusses on using Git and Visual Studio Team Services (the departments existing code version control tool of choice).
```
Comment (drmarkc13) [inline index.Rmd]: the Department's......(two changes - capitals and apostrophe)

File: index.Rmd [hunk 14]
```diff
@@ -9,4 +9,24 @@ link-citations: yes
...
\ No newline at end of file
+# Preface {-}
+
+## Overview
+
+This book aims act as a resource for analysts on how and why they should use version control.
+
+It is in the context of the Department for Education Corporate environment and focusses on using Git and Visual Studio Team Services (the departments existing code version control tool of choice). 
+
+R and R Studio are used as code base we are version controlling and IDE and the examples will be a mix of using R studio and the command line.
```
Comment (drmarkc13) [inline index.Rmd]: Unclear what this means......missing word somewhere?

should it be 'code bases'?

Define what IDE means?

File: index.Rmd [hunk 15]
```diff
@@ -9,4 +9,24 @@ link-citations: yes
...
+
+It is in the context of the Department for Education Corporate environment and focusses on using Git and Visual Studio Team Services (the departments existing code version control tool of choice). 
+
+R and R Studio are used as code base we are version controlling and IDE and the examples will be a mix of using R studio and the command line. 
+
+## Ways of Working
+
+On a high level this resource is advocating a new way of working for analytical projets which can be summarised as follows:
+
+1. **Plan your workload up front** - This will help you not only better understad the project but write less, better code to get the job done.
```
Comment (drmarkc13) [inline index.Rmd]: 'more economical code' instead of 'less code'?

Comment (drmarkc13) [inline index.Rmd]: understand

File: 01-intro.Rmd (hunk 1 above)
Comment (adamrobinson361) [reply 01-intro.Rmd]: Implemented

File: index.Rmd [hunk 16]
```diff
@@ -9,4 +9,24 @@ link-citations: yes
...
+
+2. **Take one task at a time** - Focus on one specifc item of work at a time (substantial enough that it warrants being its own task but not so large that its not clear what has change). Once its complete get it reviewed and move onto the next one. e.g. create a script that imports and cleans all of the data sources.
```
Comment (drmarkc13) [inline index.Rmd]: specific

it's (twice)

Once it's complete, get .....

File: index.Rmd [hunk 17]
```diff
@@ -9,4 +9,24 @@ link-citations: yes
...
+
+3. **Collaborate** - Distribute distinct items of work accross analysts and bring these together to build the final product and enforce code review when there is a proposal to change the main version of the code.
```
Comment (drmarkc13) [inline index.Rmd]: across

File: 01-intro.Rmd (hunk 2 above)
Comment (adamrobinson361) [reply 01-intro.Rmd]: Implemented

File: 01-intro.Rmd (hunk 3 above)
Comment (adamrobinson361) [reply 01-intro.Rmd]: Implemented

**Pull request closed without merging**